### PR TITLE
Move recursion check for zsts back to read site instead of access check site

### LINKED ISSUE #71612: Compiler panic during build of `dispatch` crate
I tried this code:

Tried to build the [`dispatch`](https://crates.io/crates/dispatch) crate.

I expected to see this happen: Successful build.
Instead, this happened: Compiler panic.

## Meta
`rustc --version --verbose`:
rustc 1.45.0-nightly (e83f75634 2020-04-26)
binary: rustc
commit-hash: e83f7563495dbe2629b0cbc738afb0808c4482e1
commit-date: 2020-04-26
host: x86_64-apple-darwin
release: 1.45.0-nightly
LLVM version: 9.0

Backtrace:
```
error: internal compiler error: src/librustc_mir/interpret/validity.rs:396: Unexpected error during ptr inbounds test: cannot read from foreign (extern) static DefId(0:40 ~ dispatch[6e69]::ffi[0]::[0]::_dispatch_queue_attr_concurrent[0])

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:904:9
stack backtrace:
   0:        0x1100d19df - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h5e3a63e7985407d7
   1:        0x11010b7ee - core::fmt::write::h26bd57c1fd6e9c98
   2:        0x1100c2977 - std::io::Write::write_fmt::h1b039c1010aa04fe
   3:        0x1100d64fa - std::panicking::default_hook::{{closure}}::h7596190ab59d75d5
   4:        0x1100d623c - std::panicking::default_hook::h8a07383816b1a388
   5:        0x114198848 - rustc_driver::report_ice::h296b79ffcf51ce3e
   6:        0x1100d6c15 - std::panicking::rust_panic_with_hook::h0bcea39f2cca3a4b
   7:        0x118234b96 - std::panicking::begin_panic::h66e22c23a1549a6e
   8:        0x117e1cfb7 - rustc_errors::HandlerInner::bug::hb332c3c5946e8455
   9:        0x117e1b9a7 - rustc_errors::Handler::bug::hd936ee9ba186a04d
  10:        0x11789ab39 - rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}::head52c037401cd39
  11:        0x11788f3b6 - rustc_middle::ty::context::tls::with_opt::{{closure}}::h1175b0b6ce73192f
  12:        0x11788f33c - rustc_middle::ty::context::tls::with_opt::hab77656c6808d61d
  13:        0x11789aa48 - rustc_middle::util::bug::opt_span_bug_fmt::he5e4627c7adae268
  14:        0x1182152eb - rustc_middle::util::bug::bug_fmt::h3d982ecec1b3b7c7
  15:        0x116b8b5d3 - rustc_mir::interpret::validity::ValidityVisitor<M>::check_safe_pointer::h291a2bee9f7f4cfe
  16:        0x116b8f9bb - rustc_mir::interpret::validity::ValidityVisitor<M>::try_visit_primitive::hb38fb3a8b177cdeb
  17:        0x116cff30f - rustc_mir::interpret::validity::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::validate_operand_internal::h0fd562892c38caa3
  18:        0x116e30f2d - rustc_mir::const_eval::eval_queries::const_eval_validated_provider::h3c9275a9f466ee97
  19:        0x116a40e8e - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::const_eval_validated>::compute::hcdb7ef1607ca94a8
  20:        0x116b6ffb8 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::hc3e13af361aba2b0
  21:        0x116a7ae95 - rustc_query_system::query::plumbing::get_query::h854993e300e4fae1
  22:        0x116e3061a - rustc_mir::const_eval::eval_queries::const_eval_validated_provider::h3c9275a9f466ee97
  23:        0x117cac8e5 - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::const_eval_validated>::compute::hcdb7ef1607ca94a8
  24:        0x117ad5158 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::h6efc840c34b85f25
  25:        0x117ca414a - rustc_query_system::query::plumbing::get_query::hff598a8d50b70459
  26:        0x1179b4411 - rustc_middle::mir::interpret::queries::<impl rustc_middle::ty::context::TyCtxt>::const_eval_poly::h4c0c73d36c3ca2d5
  27:        0x11746dd7e - <rustc_lint::BuiltinCombinedLateLintPass as rustc_lint::passes::LateLintPass>::check_item::ha3b5b1de31baab6d
  28:        0x114469423 - rustc_hir::intravisit::Visitor::visit_nested_item::h864f6dfe5b2dc4d6
  29:        0x11443397d - rustc_hir::intravisit::walk_item::h7bff3cbd983d4c68
  30:        0x11446942e - rustc_hir::intravisit::Visitor::visit_nested_item::h864f6dfe5b2dc4d6
  31:        0x11442ec9d - rustc_hir::intravisit::walk_crate::hcc8023c9ea6fe572
  32:        0x114449a77 - rustc_session::utils::<impl rustc_session::session::Session>::time::h5b6624f93e07a70e
  33:        0x11444c051 - rustc_session::utils::<impl rustc_session::session::Session>::time::hc7d512eb197226d2
  34:        0x1144a0539 - rustc_interface::passes::analysis::{{closure}}::{{closure}}::h84231c11c9470877
  35:        0x114448e8a - rustc_session::utils::<impl rustc_session::session::Session>::time::h3e5c6d62075716bf
  36:        0x1144a042e - rustc_interface::passes::analysis::h4a4e1a01585c6f70
  37:        0x1141acb5a - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute::h5be5c3012b981eba
  38:        0x1142ea159 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::hd018b5d7bb8cf53c
  39:        0x1141ad888 - rustc_query_system::query::plumbing::get_query::h1a03c197423e6671
  40:        0x11432075c - rustc_middle::ty::context::tls::enter_global::h7257fd220d03bcfb
  41:        0x1142f2fe1 - rustc_interface::interface::run_compiler_in_existing_thread_pool::hb089841313bdad7a
  42:        0x1141a0ed9 - scoped_tls::ScopedKey<T>::set::hd33ab82953a5937b
  43:        0x11419dfa5 - rustc_ast::attr::with_globals::hcb8a41b5f37a8602
  44:        0x1141a692b - std::sys_common::backtrace::__rust_begin_short_backtrace::hacb8026ff037f54c
  45:        0x1142f4ffc - core::ops::function::FnOnce::call_once{{vtable.shim}}::h8428325df99f3681
  46:        0x1100e524d - std::sys::unix::thread::Thread::new::thread_start::h41525d65c59f0a79
  47:     0x7fff6a6f5e65 - _ZL12preoptimized

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.45.0-nightly (e83f75634 2020-04-26) running on x86_64-apple-darwin

note: compiler flags: -C opt-level=3 -C bitcode-in-rlib=no -C codegen-units=1 -C debuginfo=2 -C overflow-checks=on --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [const_eval_validated] const-evaluating + checking `ffi::DISPATCH_QUEUE_CONCURRENT`
#1 [const_eval_validated] const-evaluating + checking `ffi::DISPATCH_QUEUE_CONCURRENT`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error

error: could not compile `dispatch`.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

### LINKED ISSUE #71709: Internal Compiler Error in CTFE engine
```
thread 'rustc' panicked at 'places should be checked on creation: InterpErrorInfo { kind: can't access mutable globals in ConstProp, backtrace: None }', src/librustc_mir/interpret/operand.rs:237:25
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1069
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1532
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:481
  12: rust_begin_unwind
             at src/libstd/panicking.rs:385
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:89
  14: core::option::expect_none_failed
             at src/libcore/option.rs:1272
  15: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::try_read_immediate
  16: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_rvalue_into_place
  17: <rustc_mir::transform::const_prop::ConstPropagator as rustc_middle::mir::visit::MutVisitor>::visit_statement
  18: <rustc_mir::transform::const_prop::ConstPropagator as rustc_middle::mir::visit::MutVisitor>::visit_body
  19: <rustc_mir::transform::const_prop::ConstProp as rustc_mir::transform::MirPass>::run_pass
  20: rustc_mir::transform::run_passes
  21: rustc_mir::transform::run_optimization_passes
  22: rustc_mir::transform::optimized_mir
  23: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  24: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  25: rustc_query_system::query::plumbing::get_query
  26: rustc_mir::monomorphize::collector::collect_items_rec
  27: rustc_mir::monomorphize::collector::collect_items_rec
  28: rustc_mir::monomorphize::collector::collect_items_rec
  29: rustc_mir::monomorphize::collector::collect_items_rec
  30: rustc_mir::monomorphize::collector::collect_items_rec
  31: rustc_mir::monomorphize::collector::collect_items_rec
  32: rustc_mir::monomorphize::collector::collect_items_rec
  33: rustc_mir::monomorphize::collector::collect_items_rec
  34: rustc_session::utils::<impl rustc_session::session::Session>::time
  35: rustc_mir::monomorphize::collector::collect_crate_mono_items
  36: rustc_mir::monomorphize::partitioning::collect_and_partition_mono_items
  37: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::collect_and_partition_mono_items>::compute
  38: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  39: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  40: rustc_query_system::query::plumbing::get_query
  41: rustc_codegen_ssa::base::codegen_crate
  42: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  43: rustc_interface::passes::start_codegen
  44: rustc_middle::ty::context::tls::enter_global
  45: rustc_interface::queries::Queries::ongoing_codegen
  46: rustc_interface::interface::run_compiler_in_existing_thread_pool
  47: scoped_tls::ScopedKey<T>::set
  48: rustc_ast::attr::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.45.0-nightly (fa51f810e 2020-04-29) running on x86_64-unknown-linux-gnu

note: compiler flags: -C opt-level=3 -C bitcode-in-rlib=no -C codegen-units=16 -C debuginfo=2 -C incremental -C link-arg=-fuse-ld=lld --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [optimized_mir] processing `window::glfw_window::EventsPoll::new`
#1 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack
```

The specified method, EventsPoll::new looks like:

```
static mut GLFW: Option<glfw::Glfw> = None;
pub fn new() -> Result<EventsPoll>{
        //TODO: Give the option to choose FAIL / LOG?
        let glfw = unsafe{ if let Some(glfw) = GLFW {
            glfw
        }else{
            let glfw = glfw::init(glfw::LOG_ERRORS)
                .map_err(|e| util::Error::with_cause("Error initializing glfw", e))?;
            GLFW = Some(glfw);
            glfw
        }};

        Ok(EventsPoll( Arc::new( RwLock::new(EventsPollData{
            glfw,
            should_close: 1,
        }))))
}
```

might be related with the static mut?, but haven't tried to replicate it on a smaller example, will report back if i do so

### PR DESCRIPTION
**Pull request by oli-obk**
Reverts https://github.com/rust-lang/rust/pull/71140#discussion_r413709446

Fix #71612
Fix #71709

r? @RalfJung

### CODE REVIEW

File: src/test/ui/consts/static-ice.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// check-pass
```
Comment (RalfJung) [inline src/test/ui/consts/static-ice.rs:1]: Could you pick a more descriptive file name (`extern-static-validity-ice.rs` or so) and/or add a comment? The problem is that we have a static referring to an `extern static`, and validity checking needs the static's size to make sure the reference does not dangle.

File: src/librustc_mir/interpret/operand.rs [hunk 2]
```diff
@@ -240,6 +240,15 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         {
             Some(ptr) => ptr,
             None => {
+                if let Scalar::Ptr(ptr) = mplace.ptr {
+                    // We may be reading from a static.
+                    // In order to ensure that `static FOO: Type = FOO;` causes a cycle error
+                    // instead of magically pulling *any* ZST value from the ether, we need to
+                    // actually access the referenced allocation. The caller is likely
+                    // to short-circuit on `None`, so we trigger the access here to
+                    // make sure it happens.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/operand.rs:249]: This part of the comment ("The caller is likely") makes no sense any more.

**Comment by RalfJung** [comment]
Hm... honestly I am not convinced if that is the right fix long-term. I anyway plan to change `check_offset_align` to return an `Option<&Allocation>` or so, as part of cleaning up the `Allocation` API in preparation for https://github.com/rust-lang/miri/issues/841. There's just so many other things that keep preempting this task.^^

However, there is no reason to do this re-work *now*, and this adds a testcase to make sure I find a good way to handle that when I do my planned changes. So, fine for me.

File: src/test/ui/consts/static-ice.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    static _dispatch_queue_attr_concurrent: [u8; 0];
+}
+
+static DISPATCH_QUEUE_CONCURRENT: &'static [u8; 0] =
+    unsafe { &_dispatch_queue_attr_concurrent };
+
+fn main() {
+    *DISPATCH_QUEUE_CONCURRENT;
+    new();
+}
```
Comment (RalfJung) [inline src/test/ui/consts/static-ice.rs:27]: Also maybe could you also add a testcase from https://github.com/rust-lang/rust/issues/71709? The ICE messages are quite different, so better safe than sorry in terms of test coverage...

Comment (oli-obk) [reply src/test/ui/consts/static-ice.rs:27]: This test tests both issues at the same time.

File: src/test/ui/consts/static-ice.rs (hunk 1 above)
Comment (oli-obk) [reply src/test/ui/consts/static-ice.rs:1]: Can I just refer to the two issues? The sites where the ICE occurred was fairly random.

**Comment by oli-obk** [comment]
> Hm... honestly I am not convinced if that is the right fix long-term.

Neither am I. I did try to fix it in `memory.rs` first, but that ended up looking very hacky, so I chose to go for a simple fix here and leave future refactorings to... well the future.

> I anyway plan to change `check_offset_align` to return an `Option<&Allocation>` or so, as part of cleaning up the `Allocation` API in preparation for [rust-lang/miri#841](https://github.com/rust-lang/miri/issues/841). There's just so many other things that keep preempting this task.^^
> 
> However, there is no reason to do this re-work _now_, and this adds a testcase to make sure I find a good way to handle that when I do my planned changes. So, fine for me.

File: src/test/ui/consts/static-ice.rs (hunk 1 above)
Comment (RalfJung) [reply src/test/ui/consts/static-ice.rs]: Sure (I thought it was only one of them, sorry).

File: src/test/ui/consts/ice-zst-static-access.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+// check-pass
```
Comment (RalfJung) [inline src/test/ui/consts/ice-zst-static-access.rs:1]: Are we sure "check" is enough?

**Comment by RalfJung** [comment]
r=me with that last question resolved.

File: src/test/ui/consts/ice-zst-static-access.rs (hunk 4 above)
Comment (oli-obk) [reply src/test/ui/consts/ice-zst-static-access.rs:1]: Yes, the const prop one occurred during mir optimization of `fn new`, which will still happen in check mode, and the static one happened during evaluation of the static's initializer, which also happens in check mode.

**Pull request merged**